field.class.php

```diff
@@ -461,9 +461,33 @@ public function edit_field_set_required($mform) {
             // This sets the mform field to required -- but it is not set on the db in {user_info_field} required => 1.
             $requiredifvisible = $this->field->param7;
             if ($requiredifvisible) {
+                $req = false;
+                // Warning HACK. We need to get the POSTed Data, to decide IF other data is required.
+                // This is still safe because it is all in 1 request, its just about how its treated.
+                $parentval = optional_param('profile_field_' . $this->field->param3, null, PARAM_TEXT);
+                if (!$parentval) {
+                    // We should check if this is a locked value that is already set in the DB.
+                    // In this case, we need to pull from there, it wont be in POST.
+                     // Obtain the field so we can hunt for the data.
+                    $select = $DB->sql_compare_text('shortname') . ' = ' . $DB->sql_compare_text(':shortname');
+                    $field = $DB->get_record_select('user_info_field', $select, ['shortname' => $this->field->param3]);
+                    $params = [
+                        'userid' => $USER->id,
+                        'fieldid' => $field->id,
+                    ];
+                    $infodata = $DB->get_record('user_info_data', $params);
+                    if (!$infodata->data) {
+                        // Always required in a completely blank form.
+                        $req = true;
+                    } else if ($infodata->data === $this->field->param4) {
+                        $req = true;
+                    }
+                } else if($parentval === $this->field->param4) {
+                    $req = true;
+                }
```

Comment:
Though this has been merged I wanted to do a proper review on this point:

- This should probably be moved to its own function, probably a private helper method, perhaps something like `$this->some_field_is_required()`
- `$req` should be `$required` to disambiguate it from "request or required"
- Nested if statements are hard to follow

I zoomed out a bit and saw how difficult to follow this whole function appears to be.

Reply:
I agree the logic here is tough. Its a bit easier if you have a complete grip on the plugin, but its not great. I didnt split it to its own function purely because it is already nested inside of a specific function for settting 'required' so the extra abtraction didnt feel entirely necessary.